# 1. Загрузка датасета с помощью библиотеки Pandas

In [209]:
import pandas as pd
df = pd.read_csv('./movies.csv', sep=';')

# 2. Вывод первых 20 строк с помощью метода head

In [210]:
df.head(20)

title  release_year  score  \
0      David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                     Inception        2010.0    8.8   
2                                  Forrest Gump        1994.0    8.8   
3                                    Anbe Sivam        2003.0    8.7   
4                            Bo Burnham: Inside        2021.0    8.7   
5                           Saving Private Ryan        1998.0    8.6   
6                              Django Unchained        2012.0    8.4   
7                                        Dangal        2016.0    8.4   
8                        Bo Burnham: Make Happy        2016.0    8.4   
9                         Louis C.K.: Hilarious        2010.0    8.4   
10              Dave Chappelle: Sticks & Stones        2019.0    8.4   
11                                     3 Idiots        2009.0    8.4   
12                                 Black Friday        2004.0    8.4   
13                                 Super Deluxe        2019.0    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom        2015.0    8.3   
15                  Once Upon a Time in America        1984.0    8.3   
16                                  Taxi Driver        1976.0    8.3   
17                          Like Stars on Earth        2007.0    8.3   
18                            Bo Burnham: What.        2013.0    8.3   
19                            Full Metal Jacket        1987.0    8.3   

    number_of_votes  duration   main_genre main_production  
0           31180.0        83  documentary              GB  
1         2268288.0       148        scifi              GB  
2         1994599.0       142        drama              US  
3           20595.0       160       comedy              IN  
4           44074.0        87       comedy              US  
5         1346020.0       169        drama              US  
6         1472668.0       165      western              US  
7          180247.0       161       action              IN  
8           14356.0        60       comedy              US  
9           11973.0        84       comedy              US  
10          25687.0        65       comedy              US  
11         385782.0       170       comedy              IN  
12          20611.0       143        crime              IN  
13          13680.0       176     thriller              IN  
14          17710.0        98  documentary              UA  
15         342335.0       229        drama              US  
16         795222.0       113        crime              US  
17         188234.0       165        drama              IN  
18          11488.0        60       comedy              US  
19         723306.0       116        drama              GB

# 3. Выполнить обзор данных - описать столбцы и вашу предметную область

Предметная область - набор данных о фильмах. 
Столбцы: название фильма, год выхода фильма, оценка, количество голосов, длительность фильма, жанр, производство

# 4. С помощью метода info оценить данные (есть ли пропуски, сколько всего строк, какие типы данных у столбцов)

In [211]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


Пропуски присутствуют в:
- title: 5 пропусков
- release_year: 2 пропуска
- score: 3 пропуска
- number_of_votes: 1 пропуск
- main_production: 1 пропуск

# 5. Применить describe. С помощью describe оценить числовые столбцы (если они есть).

In [212]:
df.describe()

release_year       score  number_of_votes    duration
count    387.000000  386.000000     3.880000e+02  389.000000
mean    2010.976744    7.512176     1.391523e+05  123.352185
std       10.805726    0.443870     2.365279e+05   28.304800
min     1954.000000    6.900000     1.013900e+04   28.000000
25%     2007.500000    7.100000     2.055375e+04  104.000000
50%     2014.000000    7.400000     4.606950e+04  122.000000
75%     2018.000000    7.800000     1.551585e+05  139.000000
max     2022.000000    9.000000     2.268288e+06  229.000000

Оценка числовых столбцов:
1. release year: 
    - Средний год выпуска фильмов (mean) - 2010.97
    - Диапазон лет выпуска: от 1954 до 2022
    - Большинство фильмов выпущены после 2001 года (75-й перцентиль = 2018)

2. score:
    - Средний рейтинг фильмов - 7.5
    - Все фильмы имеют достаточно высокие оценки (от 6.9 до 9)
    - Медиана (50-й перцентиль) вполне высокая - 7.4

3. number_of_votes:
    - Среднее количество голосов - 1.391523e+05
    - Значительное расхождение в количестве голосов (стандартное отклонение = 2.365279e+05)
    - Максимальное количество голосов - 2.268288e+06

4. duration:
    - Средняя продолжительность фильмов: 123.35 минут
    - Диапазон продолжительности: от 28 до 229 минут
    - Большинство фильмов длится от 104 до 139 минут

# 6. Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.

In [213]:
import re

def check_column_names(columns):
    for column in columns:
        if not re.match("^[A-Za-z0-9_]+$", column):
            print(f"Название столбца '{column} содержит неанглийские буквы или спец символы'")
        else:
            print(f"Название столбца '{column} состоит только из английских букв'")

print(df.columns)
check_column_names(df.columns)

Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')
Название столбца 'title состоит только из английских букв'
Название столбца 'release_year состоит только из английских букв'
Название столбца 'score состоит только из английских букв'
Название столбца 'number_of_votes состоит только из английских букв'
Название столбца 'duration состоит только из английских букв'
Название столбца 'main_genre состоит только из английских букв'
Название столбца 'main_production состоит только из английских букв'


Проблемы не обнаружены.
Пояснения:
1. Названия столбцов написаны в стиле snake_case, что является общепринятой практикой наименования переменных и объектов в Python
2. Названия столбцов понятны и точно описывают содержимое столбца
3. Отсутствуют опечатки, пробелы и прочие лишние символы в наименованиях столбцов
4. Наименования столбцов состоят исключительно из английских символов.

# 7. Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением). Не забудьте пояснения!

In [214]:
print(df.isna().sum())

title              5
release_year       2
score              3
number_of_votes    1
duration           0
main_genre         0
main_production    1
dtype: int64


Пропуски: 
- 5 Наименований: данные строки подлежат удалению, т.к. без названия фильма эти строки теряют смысл
- Остальные пропуски: будет использоваться замена данных
    - release_year заменяется на среднее значение (2010.97)
    - score заменяется на среднее значение (7.5)

In [215]:
df = df.dropna(subset=['title'])

print(df.isna().sum())
print("\n")

df['release_year'] = df['release_year'].fillna(2010.97)
df['score'] = df['score'].fillna(7.5)

print(df.isna().sum())

title              0
release_year       1
score              1
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64


title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64


# 8. Проверьте данные на наличие явных и неявных дубликатов. Удалите дубликаты, если они есть.

Поиск и удаление явных дубликатов

In [216]:
df[df.duplicated()]
print(f"Количество дубликатов: {df.duplicated().sum()}")

df = df.drop_duplicates().reset_index()
df = df.drop_duplicates().reset_index(drop=True)

df[df.duplicated()]
print(f"Количество дубликатов: {df.duplicated().sum()}")

Количество дубликатов: 2
Количество дубликатов: 0


Поиск неявных дубликатов:

In [200]:
print(f"Количество уникальных жанров: {df['main_genre'].unique().size}\n")
print(df['main_genre'].unique())

Количество уникальных жанров: 17

['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'COMEDY' 'dramaa' 'musical']


Неявные дубликаты: 
- Ошибка в слове "dramaa" (лишняя a)
- COMEDY - нужно заменить на нижний регистр
- scifi заменяется на sci-fi

In [201]:
df['main_genre'] = df['main_genre'].replace('dramaa', 'drama')
df['main_genre'] = df['main_genre'].str.lower() # Для приведения COMEDY к нижнему регистру, остальные жанры уже в lowercase
df['main_genre'] = df['main_genre'].replace('scifi', 'sci-fi')

print(f"Количество уникальных жанров: {df['main_genre'].unique().size}\n")
print(df['main_genre'].unique())

Количество уникальных жанров: 15

['documentary' 'sci-fi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'musical']


# 9. Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.

In [202]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 382 entries, 0 to 381
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   index            382 non-null    int64  
 1   title            382 non-null    object 
 2   release_year     382 non-null    float64
 3   score            382 non-null    float64
 4   number_of_votes  382 non-null    float64
 5   duration         382 non-null    int64  
 6   main_genre       382 non-null    object 
 7   main_production  382 non-null    object 
dtypes: float64(3), int64(2), object(3)
memory usage: 24.0+ KB


release_year типа float64 меняем на int64
number_of_votes типа float64 меняем на int64

In [203]:
df['release_year'] = df['release_year'].astype('int64')
df['number_of_votes'] = df['number_of_votes'].astype('int64')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 382 entries, 0 to 381
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   index            382 non-null    int64  
 1   title            382 non-null    object 
 2   release_year     382 non-null    int64  
 3   score            382 non-null    float64
 4   number_of_votes  382 non-null    int64  
 5   duration         382 non-null    int64  
 6   main_genre       382 non-null    object 
 7   main_production  382 non-null    object 
dtypes: float64(1), int64(4), object(3)
memory usage: 24.0+ KB


# 10. Осуществите группировки и создайте сводные таблицы в соответствии с вариантом.

Задание 1 : Группировка - год выхода и количество фильмов каждого жанра (main_genre)

In [204]:
print(df.groupby(['release_year', 'main_genre'])['index'].count())

release_year  main_genre
1954          romance       1
1961          war           1
1964          drama         1
1966          western       1
1967          drama         1
                           ..
2021          thriller      2
2022          comedy        1
              crime         1
              drama         2
              romance       1
Name: index, Length: 165, dtype: int64


2. Группировка - год выхода и количество фильмов каждого жанра (main_genre). Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count”. 

In [205]:
# Группировка - год выхода и количество фильмов каждого жанра (main_genre) и создание датафрейма
df_task2 = df.groupby(['release_year', 'main_genre'])['index'].count().reset_index()

# Переименование столбца
df_task2 = df_task2.rename(columns={'index': 'count'})

# Сортировка по убыванию столбца "count"
df_task2 = df_task2.sort_values(by='count', ascending=False)

print(df_task2)

     release_year main_genre  count
122          2017      drama     13
140          2019      drama     13
148          2020      drama      9
104          2015      drama      9
156          2021      drama      9
..            ...        ...    ...
46           2004     comedy      1
45           2003   thriller      1
43           2003      crime      1
42           2003     comedy      1
164          2022    romance      1

[165 rows x 3 columns]


3. Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по жанрам. Отсортировать по убыванию рейтинга. Округлить до двух знаков.

In [206]:
# Создание сводной таблицы с приведенным рейтингом фильмов по жанрам. Берём среднее значение
pivot_table = pd.pivot_table(df, values='score', index='main_genre', aggfunc='mean')

# Сортировка по убыванию рейтинга
pivot_table = pivot_table.sort_values(by='score', ascending=False)

# Округление до 2 знаков после запятой
pivot_table = pivot_table.round(2)

print(pivot_table)

             score
main_genre        
war           7.80
sci-fi        7.77
documentary   7.76
action        7.70
comedy        7.61
western       7.56
fantasy       7.48
drama         7.48
crime         7.47
thriller      7.47
romance       7.40
sports        7.40
horror        7.35
animation     7.20
musical       7.00


4. Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по жанрам - строки и годам - столбцы. Отсортировать по возрастанию main_genre.

In [207]:
# Создание сводной таблицы с рейтингом фильмов по жанрам и годам
pivot_table2 = pd.pivot_table(df, values='score', index='main_genre', columns='release_year', aggfunc='mean')

# Сортируем по возрастанию main_genre
pivot_table2 = pivot_table2.sort_index(ascending=True)

# Округление до 2 знаков после запятой
pivot_table2 = pivot_table2.round(2)

print(pivot_table2)

release_year  1954  1961  1964  1966  1967  1971  1973  1975  1976  1979  ...  \
main_genre                                                                ...   
action         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
animation      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
comedy         NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.2   NaN   7.8  ...   
crime          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.3   NaN  ...   
documentary    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
drama          NaN   NaN   7.8   NaN   7.7   NaN   NaN   NaN   NaN   NaN  ...   
fantasy        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
horror         NaN   NaN   NaN   NaN   NaN   NaN   8.1   NaN   NaN   NaN  ...   
musical        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
romance        7.5   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
sci-fi         NaN   NaN   N